### PR TITLE
WDPD-43 iDEAL payment method

### PR DESCRIPTION
**Pull request by ps-amehic**
### This PR

* Adds iDEAL payment method

### Notes

* Select "ING" as a bank provider on 3rd step in the checkout process (that is the only option that works in test environment).
* Setting of Sepa Credit Transfer config is moved to TransactionTabPostProcessing.php. This refactoring needed to be done, because we would have duplicate code for Sofort, iDEAL and any other payment method that might come in future and use the same functionality.

### Jira Links

* https://jira.parkside.at/browse/WDPD-43

### CODE REVIEW

File: Model/IdealPaymentMethod.php [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    /**
+     * @inheritdoc
+     *
+     * @param string $sAction
+     *
+     * @return SepaCreditTransferPaymentMethod
+     *
+     * @since 1.2.0
+     */
+    public function getPostProcessingPaymentMethod($sAction = '')
```
Comment (ps-mstifter) [inline Model/IdealPaymentMethod.php]: I don't think we need the default value at this point. `wdpd-41-eps-payment-method` also does not have it.

File: Model/IdealPaymentMethod.php [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+     *
+     * @since 1.2.0
+     */
+    protected static $_sName = "ideal";
+
+    /**
+     *
+     * @since 1.2.0
+     */
+    private static $_aBankOptions = [];
```
Comment (ps-cgrach) [inline Model/IdealPaymentMethod.php]: no need for `static` if you always assign it in the constructor.

File: Tests/Unit/Model/IdealPaymentMethodTest.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+use Wirecard\PaymentSdk\Transaction\IdealTransaction;
+
+use OxidEsales\Eshop\Core\Registry;
+
+class IdealPaymentMethodTest extends OxidEsales\TestingLibrary\UnitTestCase
+{
+    /**
+     * @var IdealPaymentMethod
+     */
+    private $oPaymentMethod;
```
Comment (ps-cgrach) [inline Tests/Unit/Model/IdealPaymentMethodTest.php]: can you name it $_oPaymentMethod with an underscore to have our guidelines applied!?

File: Model/IdealPaymentMethod.php [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+use Wirecard\PaymentSdk\Config\PaymentMethodConfig;
+use Wirecard\PaymentSdk\Entity\IdealBic;
+use Wirecard\PaymentSdk\Transaction\IdealTransaction;
+use Wirecard\PaymentSdk\Transaction\Transaction;
+
+use Wirecard\Oxid\Core\Helper;
+
+use OxidEsales\Eshop\Core\Registry;
+
+//use ReflectionClass;
```
Comment (ps-dfritzsch) [inline Model/IdealPaymentMethod.php]: Please **remove comment** if not needed

File: Model/IdealPaymentMethod.php [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+     * @since 1.2.0
+     */
+    public function getTransaction()
+    {
+        $oTransaction = new IdealTransaction();
+
+        $oSession = Registry::getConfig()->getSession();
+        $aDynvalues = $oSession->getVariable('dynvalue');
+
+        $oTransaction->setBic($this->_aBankOptions[$aDynvalues['bank']]);
```
Comment (ps-holler) [inline Model/IdealPaymentMethod.php]: I think this should be done in `addMandatoryTransactionData`.

File: Tests/Unit/Model/IdealPaymentMethodTest.php [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        $aDynArray = [
+            'bank' => 'INGBNL2A',
+        ];
+        Registry::getSession()->setVariable('dynvalue', $aDynArray);
+        $this->assertInstanceOf(IdealTransaction::class, $this->_oPaymentMethod->getTransaction());
+    }
+
+    /**
+     * @dataProvider getNameProvider
```
Comment (ps-holler) [inline Tests/Unit/Model/IdealPaymentMethodTest.php]: Why does this need a provider if you only test a single data set?

File: Tests/Unit/Model/IdealPaymentMethodTest.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        return [
+            'correct payment name' => ['ideal'],
+        ];
+    }
+
+    public function testGetConfigFields()
+    {
+        $aConfigFields = $this->_oPaymentMethod->getConfigFields();
+        $this->assertArrayHasKey('additionalInfo', $aConfigFields);
```
Comment (ps-holler) [inline Tests/Unit/Model/IdealPaymentMethodTest.php]: So far we always checked for the field names specifically, see: https://github.com/wirecard/oxid-ee/blob/master/Tests/Unit/Model/PaypalPaymentMethodTest.php#L69

File: Tests/Unit/Model/IdealPaymentMethodTest.php [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public function testGetCheckoutFields()
+    {
+        $aFields = $this->_oPaymentMethod->getCheckoutFields();
+        $this->assertEquals(array_keys($aFields), [
```
Comment (ps-holler) [inline Tests/Unit/Model/IdealPaymentMethodTest.php]: `assertArrayHasKey` would be easier for this case.

File: Tests/Unit/Model/IdealPaymentMethodTest.php [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    public function testGetBanks()
+    {
+        $aBanks = $this->_oPaymentMethod->getBanks();
+        $this->assertNotNull($aBanks);
+    }
+
+    public function testGetPublicFieldNames()
+    {
+        $aFieldNames = $this->_oPaymentMethod->getPublicFieldNames();
+        $this->assertNotNull($aFieldNames);
```
Comment (ps-holler) [inline Tests/Unit/Model/IdealPaymentMethodTest.php]: Same as above – check for all field names.

File: Tests/Unit/Model/IdealPaymentMethodTest.php (hunk 8 above)
Comment (nickstamat) [reply Tests/Unit/Model/IdealPaymentMethodTest.php]: @ps-amehic Please respond before resolving (also applies to the rest of the comments in this PR) 🙂

File: Model/IdealPaymentMethod.php (hunk 5 above)
Comment (ps-amehic) [reply Model/IdealPaymentMethod.php]: Moved to addMandatoryTransactionData. :+1:

File: Tests/Unit/Model/IdealPaymentMethodTest.php (hunk 6 above)
Comment (ps-amehic) [reply Tests/Unit/Model/IdealPaymentMethodTest.php]: Added an additional data to the provider. :+1:

File: Tests/Unit/Model/IdealPaymentMethodTest.php (hunk 7 above)
Comment (ps-amehic) [reply Tests/Unit/Model/IdealPaymentMethodTest.php]: :+1:

File: Tests/Unit/Model/IdealPaymentMethodTest.php (hunk 9 above)
Comment (ps-amehic) [reply Tests/Unit/Model/IdealPaymentMethodTest.php]: :+1:

File: Model/IdealPaymentMethod.php (hunk 1 above)
Comment (ps-amehic) [reply Model/IdealPaymentMethod.php]: :+1:

File: Model/IdealPaymentMethod.php (hunk 2 above)
Comment (ps-amehic) [reply Model/IdealPaymentMethod.php]: :+1:

File: Tests/Unit/Model/IdealPaymentMethodTest.php (hunk 3 above)
Comment (ps-amehic) [reply Tests/Unit/Model/IdealPaymentMethodTest.php]: :+1:

File: Model/IdealPaymentMethod.php (hunk 4 above)
Comment (ps-amehic) [reply Model/IdealPaymentMethod.php]: :+1:

File: Tests/Unit/Model/IdealPaymentMethodTest.php (hunk 8 above)
Comment (ps-amehic) [reply Tests/Unit/Model/IdealPaymentMethodTest.php]: I resolved this comment previously by accident. Now it is done (assertEquals replaced by assertArrayHasKey). :+1:

File: Controller/Admin/Transaction/TransactionTabPostProcessing.php [hunk 10]
```diff
@@ -430,12 +434,22 @@ private static function _getResultMessageArray($sType, $sMessage)
...
-            $oConfig = $this->_getPaymentMethodConfig();
-            $this->_oTransactionHandler = new TransactionHandler($this->_getBackendService($oConfig));
+        if ($this->_oTransactionHandler) {
+            return $this->_oTransactionHandler;
         }
 
-        return $this->_oTransactionHandler;
+        $oConfig = $this->_getPaymentMethodConfig();
+
+        $oSepaCtPayment = PaymentMethodHelper::getPaymentById(SepaCreditTransferPaymentMethod::getName(true));
```
Comment (tomazpu) [inline Controller/Admin/Transaction/TransactionTabPostProcessing.php]: This is in my opinion bad practice. As now the SEPA CT config is added all the time. Also this code is not belonging here, here the only thing that should be is $oConfig = $this->_getPaymentMethodConfig(); all other stuff needs to happen in payment methods. Please solve this issue with inheritance or a different class structure. Maybe we can solve it with adding a interface or with another technique.  Then we can have sperate implementations of one where only payment method Config is loaded and one where we load a SEPA CT config if its a refund. If you need help please contact me, but i think with this we just shifting the responsibility to a wrong place.

Comment (ps-amaureder) [reply Controller/Admin/Transaction/TransactionTabPostProcessing.php]: I changed it, so all payment methods which need sepa ct for refund are inherited from sepa ct

File: Model/IdealPaymentMethod.php [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    protected static $_sName = "ideal";
+
+    /**
+     * @inheritdoc
+     *
+     * @var bool
+     *
+     * @since 1.2.0
+     */
+    protected static $_bMerchantOnly = false;
```
Comment (ps-holler) [inline Model/IdealPaymentMethod.php]: Since most payment methods are not merchant only we might want to set the default to `false`.

Comment (ps-amaureder) [reply Model/IdealPaymentMethod.php]: we do, but this method inherits from sepa ct and there it is set to true

Comment (ps-holler) [reply Model/IdealPaymentMethod.php]: Nevermind, I was under the impression that Sepa CT was an abstract class.

File: Controller/Admin/Transaction/TransactionTabPostProcessing.php [hunk 12]
```diff
@@ -13,14 +13,16 @@
 use OxidEsales\Eshop\Core\Exception\StandardException;
 
 use Wirecard\Oxid\Core\Helper;
+use Wirecard\Oxid\Core\PaymentMethodHelper;
 use Wirecard\Oxid\Core\TransactionHandler;
 use Wirecard\Oxid\Core\PaymentMethodFactory;
+use Wirecard\Oxid\Extend\Model\Payment;
```
Comment (tomazpu) [inline Controller/Admin/Transaction/TransactionTabPostProcessing.php]: I don't see any use of this class

File: Controller/Admin/Transaction/TransactionTabPostProcessing.php [hunk 13]
```diff
@@ -338,10 +340,14 @@ protected function _getPostProcessingActions()
      */
     private function _filterPostProcessingActions($aPossibleOperations, $oPaymentMethod)
     {
-        if ($oPaymentMethod->getName(true) === SofortPaymentMethod::getName(true)
-            && !$oPaymentMethod->getPostProcessingPaymentMethod(Transaction::ACTION_CREDIT)
-                ->getPayment()->oxpayments__oxactive->value) {
-            return [];
+        $oTransaction = $oPaymentMethod->getPostProcessingTransaction(Transaction::ACTION_CREDIT, $this->oTransaction);
```
Comment (tomazpu) [inline Controller/Admin/Transaction/TransactionTabPostProcessing.php]: Why is the ACTION_CREDIT hard coded? Is it not depending on what we are doing?

File: Model/SepaCreditTransferPaymentMethod.php [hunk 14]
```diff
@@ -71,19 +74,24 @@ public function getTransaction()
...
-     * @since 1.1.0
+     * @return SepaCreditTransferTransaction
+     *
+     * @since 1.2.0
      */
-    public function addPostProcessingTransactionData(&$oTransaction, $oParentTransaction)
+    public function getPostProcessingTransaction($sAction, $oParentTransaction)
     {
+        $oTransaction = new SepaCreditTransferTransaction();
+
```
Comment (tomazpu) [inline Model/SepaCreditTransferPaymentMethod.php]: Is the Mandate also needed for post processing operations? is it strictly written in the user story?

File: Controller/Admin/Transaction/TransactionTabPostProcessing.php (hunk 12 above)
Comment (ps-amaureder) [reply Controller/Admin/Transaction/TransactionTabPostProcessing.php]: You are right, I removed it and we will adapt our code sniffer

File: Controller/Admin/Transaction/TransactionTabPostProcessing.php (hunk 13 above)
Comment (ps-amaureder) [reply Controller/Admin/Transaction/TransactionTabPostProcessing.php]: fixed

File: Model/SepaCreditTransferPaymentMethod.php (hunk 14 above)
Comment (ps-amaureder) [reply Model/SepaCreditTransferPaymentMethod.php]: there is no user story for sepa ct.
I followed the payment sdk examples where mandate is provided to sepa ct
However I tried it without a mandate and it worked, so I will change it

**Pull request merged**